server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java

```diff
@@ -153,11 +153,44 @@ default boolean supportsConcurrentMultipartUploads() {
         return false;
     }
 
+    /**
+     * Provides an {@link InputStream} to read a part of the blob content.
+     */
+    @FunctionalInterface
+    interface BlobMultiPartInputStreamProvider extends CheckedBiFunction<Long, Long, InputStream, IOException> {
+        /**
+         * Provides an {@link InputStream} to read a part of the blob content.
+         *
+         * @param offset        the offset in the blob content to start reading bytes from
+         * @param length        the number of bytes to read
+         * @return              an {@link InputStream} to read a part of the blob content.
+         * @throws IOException  if something goes wrong opening the input stream
+         */
+        @Override
+        InputStream apply(Long offset, Long length) throws IOException;
```

Comment:
Can we make the longs primitive to avoid boxing and thinking about nulls?

```suggestion
        InputStream apply(long offset, long length) throws IOException;
```